File: SwiftySheetsDemo/Parser/AST/ASTNumericLiteral.m
//
//  ASTNumericLiteral.m
//  SwiftySheetsDemo
//
//  Created by imihaly on 09/07/16.
//  Copyright © 2016 Agnes Vasarhelyi. All rights reserved.
//

#import "ASTNumericLiteral.h"

@interface ASTNumericLiteral ()

@property(nonatomic, copy, readwrite) NSNumber *value;

@end

@implementation ASTNumericLiteral

+ (instancetype)literalWithFloat:(float)value {
	return [[self alloc] initWithFloat:value];
}

+ (instancetype)literalWithDecimalString:(NSString *)decimalString {
	return [[self alloc] initWithDecimalString:decimalString];
}

+ (instancetype)literalWithHexaString:(NSString *)hexaString {
	return [[self alloc] initWithHexaString:hexaString];
}

- (instancetype)initWithFloat:(float)value {
	self = [super init];
	if(self) {
		self.value = @(value);
	}
	return self;
}

- (instancetype)initWithDecimalString:(NSString *)decimalString {
	self = [super init];
	if(self) {
		self.value = [NSNumber numberWithFloat:decimalString.floatValue];
	}
	return self;
}

- (instancetype)initWithHexaString:(NSString *)hexaString {
	self = [super init];
	if(self) {
		unsigned result = 0;
		NSScanner *scanner = [NSScanner scannerWithString:hexaString];
		
		[scanner setScanLocation:1]; // bypass '#' character
		if([scanner scanHexInt:&result]) {
			self.value = @(result);
		} else {
			@throw [NSError errorWithDomain:@"ParserErrorDomain"
									   code:-2
								   userInfo:@{NSLocalizedDescriptionKey :  [NSString stringWithFormat:@"Failed to parse hexadecimal string: %@", hexaString]}];
		}
	}
	return self;	
}

- (ASTExprType)type {
	return ASTExprType_Numeric;
}

- (id)eval {
	return self.value;
}

- (NSString *)description {
	return [NSString stringWithFormat:@"Num<%@>", self.value];
}

@end

File: SwiftySheetsDemo/Parser/AST/ASTNumericLiteral.swift
//
//  ASTNumericLiteral.swift
//  SwiftySheetsDemo
//
//  Created by Alföldi Norbert on 19/07/16.
//  Copyright © 2016 Agnes Vasarhelyi. All rights reserved.
//

import Foundation

@objc public class ASTNumericLiteral : NSObject, ASTExpression, FloatLiteralConvertible, StringLiteralConvertible {
	public var value: Double
	
	public required init(floatLiteral value: FloatLiteralType) {
		self.value = value
	}
	
	public required init(stringLiteral value: StringLiteralType) {
		self.value = ASTNumericLiteral.convert(value) ?? 0
	}
	
	public typealias UnicodeScalarLiteralType = StringLiteralType
	public required init(unicodeScalarLiteral value: UnicodeScalarLiteralType) {
		self.value = ASTNumericLiteral.convert(value) ?? 0
	}
	
	public typealias ExtendedGraphemeClusterLiteralType = StringLiteralType
	public required init(extendedGraphemeClusterLiteral value: ExtendedGraphemeClusterLiteralType) {
		self.value = ASTNumericLiteral.convert(value) ?? 0
	}
	
	private static func convert(value: String) -> Double? {
		if value.hasPrefix("#") {
			var intValue: CUnsignedInt = 0
			let scanner = NSScanner(string: value.substringFromIndex(value.startIndex.advancedBy(1)))
			scanner.scanHexInt(&intValue)
			
			return Double(intValue)
		} else {
			return Double(value)
		}
	}
	
	public func eval() -> AnyObject {
		return self.value
	}
}
